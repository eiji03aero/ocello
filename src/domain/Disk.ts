export enum DiskColors {
  Black = "Black",
  White = "White",
}

export interface DiskArgs {
  color: DiskColors;
}

export class Disk {
  color: DiskColors;

  constructor (args: DiskArgs) {
    this.color = args.color;
  }

  static Black () {
    return new Disk({ color: DiskColors.Black });
  }

  static White () {
    return new Disk({ color: DiskColors.White });
  }
}
