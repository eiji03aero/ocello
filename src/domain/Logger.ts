export enum LogLevel {
  Log = "Log",
  Warn = "Warn",
  Error = "Warn",
}

export class Log {
  text: string;
  level: LogLevel;
  createdAt: string;

  constructor (args: { text: string, level: LogLevel }) {
    this.text = args.text;
    this.level = args.level;
    const now = new Date();
    this.createdAt = `${now.getHours()}:${now.getMinutes()}`;
  }
}

export interface LoggerArgs {
  logs: Log[];
}

export class Logger {
  logs: Log[];

  static Blank () { return new this({ logs: [] as Log[] }); }

  constructor (args: LoggerArgs) {
    this.logs = args.logs ;
  }

  log (text: string) {
    const log = new Log({ text: text, level: LogLevel.Log });
    this.logs.push(log);
  }

  warn (text: string) {
    const log = new Log({ text: text, level: LogLevel.Warn });
    this.logs.push(log);
  }

  error (text: string) {
    const log = new Log({ text: text, level: LogLevel.Error });
    this.logs.push(log);
  }
}
