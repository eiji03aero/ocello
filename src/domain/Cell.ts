import { Disk, DiskColors } from "./Disk";

export enum CellState {
  Blank = "Blank",
  Placed = "Placed",
}

interface CellArgs {
  state: CellState;
  disk?: Disk;
}

export class Cell {
  state: CellState;
  disk: Disk | null;

  constructor (args: CellArgs) {
    this.state = args.state || CellState.Blank;
    this.disk = args.disk || null;
  }

  static Blank () {
    return new Cell({ state: CellState.Blank });
  }

  static Black () {
    return new Cell({ state: CellState.Placed, disk: Disk.Black() });
  }

  static White () {
    return new Cell({ state: CellState.Placed, disk: Disk.White() });
  }

  static fromColor (color: DiskColors) {
    return new Cell({ state: CellState.Placed, disk: new Disk({ color: color })});
  }

  get color () {
    return this.state === CellState.Placed
      ? this.disk.color
      : '';
  }

  get isBlank () {
    return this.state === CellState.Blank;
  }

  get isPlaced () {
    return this.state === CellState.Placed;
  }
}
