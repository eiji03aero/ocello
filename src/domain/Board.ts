import { Cell } from './Cell';

export type BoardCells = Cell[][];

export type DiskCoordinates = [number, number];

export interface BoardArgs {
  data: BoardCells;
}

export class Board {
  data: BoardCells;

  constructor (args: BoardArgs) {
    this.data = args.data;
  }

  static Blank () { return new Board({ data: this.defaultData }); }
  static Random () { return new Board({ data: this.randomData }); }

  private static get defaultData () {
    return new Array(8).fill(0).map(row => {
      return new Array(8).fill(0).map(col => {
        return Cell.Blank();
      });
    });
  }

  private static get randomData () {
    return new Array(8).fill(0).map(row => {
      return new Array(8).fill(0).map(col => {
        const rand = Math.floor(Math.random() * 3);
        return (
          rand === 0 ? Cell.Blank() :
          rand === 1 ? Cell.Black() :
          Cell.White()
        );
      });
    });
  }
}
