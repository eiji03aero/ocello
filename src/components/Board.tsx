import * as React from 'react';
import * as _ from "lodash";
import styled from "styled-components";

import { Row } from "./Row";

import { Board as BoardClass, BoardRow } from "../domain/Board";
import { colors } from "../utils/colors";

const BoardStyled = styled.div`
  display: flex;
  flex-direction: column;
  border: 1px solid ${colors.black};

  @media only screen and (max-width: 767px) {
    width: 95vw;
    height: 95vw;
  }

  @media only screen and (min-width: 768px) {
    width: 50vw;
    height: 50vw;
  }
`;

export interface Props {
  board: BoardClass;
}

export const Board: React.SFC<Props> = ({
  board,
} : Props) => {
  return (
    <BoardStyled>
      { _.map(board.cells, (row: BoardRow, idx: number) => (
        <Row
          key={idx}
          rowIndex={idx}
          row={row}
        />
      ))}
    </BoardStyled>
  );
}
