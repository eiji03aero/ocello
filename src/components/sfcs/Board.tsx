import * as React from 'react';
import * as _ from "lodash";
import styled from "styled-components";

import { Row } from "./Row";

import { Board as BoardClass, BoardRow } from "../../domain/Board";
import { colors } from "../../utils/colors";

const BoardStyled = styled.div`
  display: inline-block;
  border: 1px solid ${colors.black};
`;

export interface Props {
  board: BoardClass;
}

export const Board: React.SFC<Props> = ({
  board,
} : Props) => {
  return (
    <BoardStyled>
      { _.map(board.cells, (row: BoardRow, idx: number) => (
        <Row
          key={idx}
          rowIndex={idx}
          row={row}
        />
      ))}
    </BoardStyled>
  );
}
