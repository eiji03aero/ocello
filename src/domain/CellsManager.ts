import * as _ from "lodash";
import { CellCoordinates, BoardCells } from "./Board";

export class CellManager {
  static LowerEnd = 0;
  static HigherEnd = 7;

  static getAround (coords: CellCoordinates): CellCoordinates[] {
    const seedCoords = [
      [-1, -1], [0, -1], [1, -1],
      [-1, 0], [1, 0],
      [-1, 1], [0, 1], [1, 1],
    ];
    const candidateCoords = _.map(seedCoords, (sc: CellCoordinates) => {
      const rc = coords[0] + sc[0];
      const cc = coords[1] + sc[1];
      return [rc, cc] as CellCoordinates;
    });
    return _.filter(candidateCoords, (cand: CellCoordinates) => {
      return (
        CellManager.LowerEnd <= cand[0] &&
        CellManager.LowerEnd <= cand[1] &&
        CellManager.HigherEnd >= cand[0] &&
        CellManager.HigherEnd >= cand[1]
      );
    });
  }

  static checkIfPlaceable (coords: CellCoordinates, data: BoardCells) {
    return true
  }
}
