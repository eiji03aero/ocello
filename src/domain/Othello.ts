import { Board } from "./Board";

export enum GameStatus {
  PLAYING,
  FINISHED,
}

export enum UserTurn {
  ONE,
  TWO,
}

export interface OthelloArgs {
  board: Board;
  gameStatus: GameStatus;
  userTurn: UserTurn;
}

export class Othello {
  board: Board;
  gameStatus: GameStatus;
  userTurn: UserTurn;

  constructor (args: OthelloArgs) {
    this.board = args.board;
    this.gameStatus = args.gameStatus;
    this.userTurn = args.userTurn;
  }

  static NewGame () {
    return new Othello({
      board: Board.Blank(),
      gameStatus: GameStatus.PLAYING,
      userTurn: UserTurn.ONE,
    });
  }

  static RandomGame () {
    return new Othello({
      board: Board.Random(),
      gameStatus: GameStatus.PLAYING,
      userTurn: UserTurn.ONE,
    });
  }
}
