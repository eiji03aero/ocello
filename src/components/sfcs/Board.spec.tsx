import * as React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";

import { Board } from "./Board";
import { Board as BoardClass } from "../../domain/Board";

describe('Board', () => {
  it('matches snapshot', () => {
    const board = BoardClass.NewGame();
    const wrapper = shallow(
      <Board
        board={board}
      />
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
