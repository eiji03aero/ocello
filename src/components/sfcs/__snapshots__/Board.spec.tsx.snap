// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board matches snapshot 1`] = `
<styled.div>
  <Component
    key="0"
    row={
      Array [
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
      ]
    }
    rowIndex={0}
  />
  <Component
    key="1"
    row={
      Array [
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
      ]
    }
    rowIndex={1}
  />
  <Component
    key="2"
    row={
      Array [
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
      ]
    }
    rowIndex={2}
  />
  <Component
    key="3"
    row={
      Array [
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": Disk {
            "color": "Black",
          },
          "state": "Placed",
        },
        Cell {
          "disk": Disk {
            "color": "White",
          },
          "state": "Placed",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
      ]
    }
    rowIndex={3}
  />
  <Component
    key="4"
    row={
      Array [
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": Disk {
            "color": "White",
          },
          "state": "Placed",
        },
        Cell {
          "disk": Disk {
            "color": "Black",
          },
          "state": "Placed",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
      ]
    }
    rowIndex={4}
  />
  <Component
    key="5"
    row={
      Array [
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
      ]
    }
    rowIndex={5}
  />
  <Component
    key="6"
    row={
      Array [
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
      ]
    }
    rowIndex={6}
  />
  <Component
    key="7"
    row={
      Array [
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
        Cell {
          "disk": null,
          "state": "Blank",
        },
      ]
    }
    rowIndex={7}
  />
</styled.div>
`;
